docs/logs/getting-started/Program.cs

```diff
@@ -32,6 +32,14 @@ public static void Main()
         });
 
         var logger = loggerFactory.CreateLogger<Program>();
-        logger.LogInformation("Hello from {name} {price}.", "tomato", 2.99);
+
+        logger.LogInformation(eventId: 123, "Hello from {name} {price}.", "tomato", 2.99);
+
+        if (logger.IsEnabled(LogLevel.Debug))
```

Comment:
We have been traditionally trying to keep getting started as simple as possible, and not show anything not really required (though recommended).

Agree this is a good practice to include EventId (.NET analyzers now throws warnings about duplicate event id etc.), but do we need to show them in the hello world? We could make it part of a subsequent "beyond getting started" topics, along with many other best practices. (like source-generation, structured logging etc).

The tracing tutorial is skipping any mention of "activity.IsAllDataRequested", in the getting started: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/getting-started-console/Program.cs#L35

Reply:
It looks like we have been having some struggle here - the real simple demo is essentially encouraging something which we don't intend developers to use in production code, and the code we do encourage people to write for production environments are in some "advanced" docs.

Ideally, I hope we would have something really simple and production ready.

The options for us:
1. Keep the simple one really simple, understanding that it is not encouraging the best practices.
2. Demonstrate production ready code in the getting-started tutorial, understanding that it is not as simple as a single line.
3. Somewhere between 1 and 2 - this PR was trying to take the compromise.

Which one do folks prefer? (I personally would vote for 2 or 3, I think 1 is something bad that we should get rid of).

Reply:
With the latest analyzers there are warnings generated using the extensions:

![image](https://user-images.githubusercontent.com/28367120/214120593-2db8c56e-ebdd-4137-94db-107615421bd9.png)

To make that happy...

```csharp
public class Program
{
    private static readonly Action<ILogger, string, double, Exception?> LogHello = LoggerMessage.Define<string, double>(
        LogLevel.Information,
        new EventId(1),
        "Hello from {Name} {Price}."); // Does generate a warning if not using pascal case

    public static void Main()
    {
        using var loggerFactory = LoggerFactory.Create(builder =>
        {
            builder.AddOpenTelemetry(options =>
            {
                options.AddConsoleExporter();
            });
        });

        var logger = loggerFactory.CreateLogger<Program>();

        LogHello(logger, "tomato", 2.99, null);
    }
}
```

Using the source generator:

```csharp
public class Program
{
    public static void Main()
    {
        using var loggerFactory = LoggerFactory.Create(builder =>
        {
            builder.AddOpenTelemetry(options =>
            {
                options.AddConsoleExporter();
            });
        });

        var logger = loggerFactory.CreateLogger<Program>();

        LogMessages.Hello(logger, "tomato", 2.99);
    }

    private static partial class LogMessages
    {
        [LoggerMessage(
            EventId = 1,
            Level = LogLevel.Information,
            Message = "Hello from {name} {price}.")] // Does not warn about pascal casing, it may in the future?
        public static partial void Hello(
            ILogger logger, string name, double price);
    }
}
```

My thoughts are...

* I would like to see all the code in the repo pass analysis. For me, that is the minimum bar for code quality.
* If we are going to show one of these patterns, the source generator is the lesser evil IMO. Don't like in the LoggerMessage version you are forced to pass the `null` exception parameter when logging.

I do agree with @cijothomas that we should seek for simple examples, but in this case it seems like .NET is pushing hard to move people away from the extensions. Just too expensive with boxings & object[] allocations.

Reply:
.NET's own logging doc is still showing the extension methods (and not LoggerMessage/Source-generator) https://learn.microsoft.com/en-us/dotnet/core/extensions/logging?tabs=command-line#create-logs

Modifying our hello world example to use LoggerMessage/SourceGen is a bit too much for hello world purposes in my opinion. Also I think it can potentially cause people to incorrectly assume that they need to change their logging code in order for it to work with Otel.
 
Maybe we can stick with comparable example as official .NET docs, and once it moves to show source-gen/loggermessage, we can also switch to that.

I agree with "somewhere between 1&2" part. 
Example:
For logging -  EventId, checking if_enabled before doing expensive log.
For tracing - checking sampling before populating expensive tags.
For metrics - show usage of TagList for >3 tags

Reply:
> Maybe we can stick with comparable example as official .NET docs
👍

> I would like to see all the code in the repo pass analysis.
Also 👍, but agree that an effort to fix our analysis may serve to confuse people as Cijo suggests i.e., make people think they need to adapt their code to make things work. Good to let these new patterns settle in a bit longer.

Reply:
> Updated the example code to use durable event identifier.

This makes sense to me given the output demonstrates we handle event ids.

> Added example code and comments to show how to handle expensive logging statements.

But what does this show off in OpenTelemetry terms? Why not just have the single logger.LogInformation?

Reply:
> > Added example code and comments to show how to handle expensive logging statements.
> 
> But what does this show off in OpenTelemetry terms? Why not just have the single logger.LogInformation?

This is not tight to OpenTelemetry .NET at all, it is a general ILogger usage pattern.

The reason I consider it here - I noticed many folks just copied the code without thinking about how expensive it is going to cost them. And here is my struggle (as always) - I don't feel OpenTelemetry .NET should be the ultimate place for people to learn how to properly use ILogger, but on the other hand I do feel bad if we provided anti-pattern examples, then people just followed them and fell into the pit of failures.

If Debug level is not enabled, one doesn't have to pay for the `SomeExpensiveCall()` eval cost:

```csharp
if (logger.IsEnabled(LogLevel.Debug))
{
    logger.LogDebug(eventId: 501, "System.Environment.Version: {version}.", SomeExpensiveCall());
}
```

One has to pay for the `SomeExpensiveCall()` eval cost regardless of whether Debug level is enabled or not:

```csharp
logger.LogDebug(eventId: 501, "System.Environment.Version: {version}.", SomeExpensiveCall());
```

Reply:
> I don't feel OpenTelemetry .NET should be the ultimate place for people to learn how to properly use ILogger

Agreed. I didn't think the example was previously demonstrating an anti-pattern. Though, I agree adding `logger.LogDebug` without a `logger.IsEnabled` check would be an anti-pattern.

Though, as I think about it more, from an OpenTelemetry perspective, adding our logging provider naturally incurs more overhead with logging. So, I can get behind reinforcing this good practice in our docs.